### ISSUE TITLE
SAML 2.0 HTTP-Artifact binding

### ISSUE THREAD

**Issue by paweles** [keycloak/keycloak#14073]
### Description

Support for the HTTP-Artifact binding in addition to the already implemented HTTP-Redirect and HTTP-POST bindings

This issue was already mentioned in the previous ticket system:
https://issues.redhat.com/browse/KEYCLOAK-831

### Motivation

SAML 2.0 HTTP-Artifact binding is a must-have feature for any integration with Polish Government IDP ([login.gov.pl](https://login.gov.pl/)). I believe that not only for this.

### Details

Described also in SP-Initiated SSO: POST/Artifact Bindings part of https://www.oasis-open.org/committees/download.php/27819/sstc-saml-tech-overview-2.0-cd-02.pdf

**Comment by limux82** [comment]
Is there any update on this topic?

**Comment by tmorin** [comment]
Hi,

We started a discussion, following a beginning of implementation -> #25745 

Best regards

**Comment by StefanCap** [comment]
@tmorin Using your branch I have been busy testing a connection to an external IDP (eHerkenning in this case) with keycloak acting as an SP. From what I can see artifact resolution works fine, however when the artifact contains a assertion with encrypted attributes parsing fails. 

From what I can see, keycloak can handle Assertions or EncryptedAssertions (Not too sure about this), but having an Assertion with an AttributeStatement that contains EncryptedAttributes causes this failure:

```
2024-06-05 10:21:38,073 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-1) Uncaught server error: java.lang.RuntimeException: PL00062: Parser : Unknown tag:EncryptedAttribute::location=Line number = 1

        at org.keycloak.saml.common.DefaultPicketLinkLogger.parserUnknownTag(DefaultPicketLinkLogger.java:360)
        at org.keycloak.saml.processing.core.parsers.saml.assertion.SAMLAttributeStatementParser.processSubElement(SAMLAttributeStatementParser.java:58)
        at org.keycloak.saml.processing.core.parsers.saml.assertion.SAMLAttributeStatementParser.processSubElement(SAMLAttributeStatementParser.java:31)
        at org.keycloak.saml.common.parsers.AbstractStaxParser.parse(AbstractStaxParser.java:97)
```

I might be missing something entirely here, but thought I'd ask the question.

**Comment by tmorin** [comment]
Hi @StefanCap,

Thank you for the feedback.

I did integration tests but without encrypted assertions.

However, a PR has been submitted (#29619). Did you try its content?

Best regards,

**Comment by tmorin** [comment]
@StefanCap,

Thank for the feedback.
I guess I spotted the issue. 
However, the Keycloak IDP for SAML might not be able to encrypt the Assertion but only signed them.
So, I tested with signed assertion only.

![image](https://github.com/keycloak/keycloak/assets/1984597/b799dbce-4b04-4a50-9944-2394e791ce9b)

Best

**Comment by sanderino666** [comment]
@tmorin I tested the latest nightly of KC 26 and the HTTP artifact binding works for DigiD (the Dutch government civilian IDP). I also tested eHerkenning (the Dutch government company IDP) but that doesn't work because the artifact response contains encrypted attributes in the artifact response. 

`org.keycloak.saml.processing.core.parsers.saml.assertion.SAMLAttributeParser#processSubElement` throws an exception when anything other than an `Attribute` is processed resulting in the entire response not parsed and a login failure.

From the eHerkenning perspective those encrypted attributes are additional attributes and need to be decrypted. 

Is there any way for Keycloak to support this? I can help with a PR and testing if needed.

**Comment by vernonmanuela** [comment]
> @tmorin I tested the latest nightly of KC 26 and the HTTP artifact binding works for DigiD (the Dutch government civilian IDP). I also tested eHerkenning (the Dutch government company IDP) but that doesn't work because the artifact response contains encrypted attributes in the artifact response.
> 
> `org.keycloak.saml.processing.core.parsers.saml.assertion.SAMLAttributeParser#processSubElement` throws an exception when anything other than an `Attribute` is processed resulting in the entire response not parsed and a login failure.
> 
> From the eHerkenning perspective those encrypted attributes are additional attributes and need to be decrypted.
> 
> Is there any way for Keycloak to support this? I can help with a PR and testing if needed.

Any updates?

**Comment by tmorin** [comment]
Hi @vernonmanuela,

I already replied previously. :) I am not a Keycloak maintainer, so I may provide wrong information.

FyI, I tested between two Keycloak instances with encryption enabled and did not encounter any issues. As far as I understand, others have also tested successfully.

If you think your change is part of the SAML standard, I believe you can suggest an implementation (PR) directly in the Keycloak codebase.

Best

**Comment by Marcinno** [comment]
Hey, 
I also tried to test it and got the same error:  Unknown tag:EncryptedAttribute

It seems that indeed, keycloak can handle EncryptedAssertions (not sure whether it expects EncryptedAttributes inside? ) but eHerkenning provider supports having EncryptedAttribute without EncryptedAssertion on top
Are there any updates or clues from your side?@sanderino666, @vernonmanuela Did you fix it?

Regards, 
Marcin

**Comment by sanderino666** [comment]
@Marcinno I checked the existing code but there is no easy way to tackle this in custom code without opening a PR here to implement actual proper support. I tried it but the result is making copies of all parser classes that are called. I eventually reached the point where I can parse the encrypted attribute but there are still questions from my side on how properly implement this (for example how to pass the private key through these parsers / decrypt in the correct spot).

I am currently in a "all work needs to be done before Christmas" state but I am willing to put in the effort after new years.
